### PR TITLE
Record: 11L Partial RoPE + LN Scale + EMA + XSA4 (val_bpb: 1.1248)

### PR DESCRIPTION
**Pull request by jfprincz**
## Record: 11L Partial RoPE + LN Scale + EMA + XSA4 (val_bpb: 1.1248)

**val_bpb: 1.1248** (sliding window, stride=64) | **15.6 MB** | 8xH100 SXM, 600s

### Progress from prior submissions

| | [PR #70](https://github.com/openai/parameter-golf/pull/70) | [PR #164](https://github.com/openai/parameter-golf/pull/164) | [PR #198](https://github.com/openai/parameter-golf/pull/198) | [PR #287](https://github.com/openai/parameter-golf/pull/287) | This | Delta vs #287 |
|---|---|---|---|---|---|---|
| **val_bpb (sliding)** | 1.1659 (s256) | 1.1524 (s256) | 1.1318 (s64) | 1.1271 (s64) | **1.1248 (s64)** | **-0.0023** |
| Layers | 9 | 9 | 11 | 11 | 11 | — |
| Params | 21.8M | 22.4M | 26.8M | 26.8M | 26.8M | — |
| Artifact | 14.9 MB | 15.4 MB | 15.7 MB | 15.5 MB | 15.6 MB | +0.1 MB |

Two new techniques on top of PR #287's 11-layer stack.

### Key additions over PR #287

| Change | Impact |
|--------|--------|
| **Partial RoPE (16 of 64 dims)** | Apply rotary embeddings to only 25% of head dimensions. Remaining dims use position-free attention, improving generalization. Zero new parameters. |
| **LN Scale** | RMSNorm outputs scaled by 1/sqrt(layer_idx+1). Damps deeper layers' contributions, stabilizing training. Zero new parameters. |

Everything else from PR #287 carries forward: 11 layers, XSA on last 4 layers, EMA (0.997), OrthoInit + muP, 3x MLP, int6 mixed quant + zstd-22, WD=0.04, SmearGate, BigramHash(2048), FA3, seq 2048, tuned Muon.

### Results

| Metric | Value |
|--------|-------|
| Pre-quant val_bpb | 1.1418 |
| Int6 roundtrip val_bpb | 1.1485 |
| **Int6 sliding val_bpb (s64)** | **1.1248** |
| Steps completed (600s cap) | 7,051 |
| Step time | 85ms |
| Model params | 26,829,913 |
| Artifact size | 15,612,308 bytes |

### Reproducibility (3 seeds)

| Seed | Steps | Sliding s64 | Artifact |
|------|-------|-------------|----------|
| **2025** | **7,051** | **1.1248** | **15,612,308** |
| 42 | 7,061 | 1.1250 | 15,528,666 |
| 1337 | 7,063 | 1.1253 | 15,639,340 |

Mean: **1.1250** | Variance: 0.0005 | Submitted: seed 2025

### Run command

```bash
NUM_LAYERS=11 BIGRAM_VOCAB_SIZE=2048 XSA_LAST_N=4 \
EMA_ENABLED=1 EMA_DECAY=0.997 SWA_ENABLED=0 \
ROPE_DIMS=16 LN_SCALE=1 LATE_QAT=1 QAT_THRESHOLD=0.1 \
MUON_WD=0.04 ADAM_WD=0.04 \
MATRIX_LR=0.025 SCALAR_LR=0.025 TIED_EMBED_LR=0.035 \
MUON_MOMENTUM=0.99 MUON_MOMENTUM_WARMUP_START=0.92 \
MUON_MOMENTUM_WARMUP_STEPS=1500 WARMDOWN_ITERS=3000 \
ITERATIONS=9000 MAX_WALLCLOCK_SECONDS=600 EVAL_STRIDE=64 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### Note on Late QAT

The submitted code includes a Late QAT flag (`LATE_QAT=1`) intended to enable STE int6 fake-quantization in the final 4% of training. Post-submission analysis (credit: @152334H) revealed that `torch.compile` constant-folds the `CastedLinear._qat_enabled` class attribute at first trace, so the STE branch is dead-code-eliminated and never activates during training. Late QAT had no effect on the results. The score is driven entirely by Partial RoPE and LN Scale.

### CODE REVIEW

File: records/track_10min_16mb/2026-03-21_11L_XSA4_EMA_PartialRoPE_LateQAT_1.1248/train_gpt.py [hunk 1]
```diff
@@ -0,0 +1,1588 @@
...
+    def forward(self, x: Tensor) -> Tensor:
+        return F.rms_norm(x, (x.size(-1),), eps=self.eps)
+
+
+class CastedLinear(nn.Linear):
+    _qat_enabled: bool = False
+
+    def forward(self, x: Tensor) -> Tensor:
+        w = self.weight.to(x.dtype)
+        if CastedLinear._qat_enabled and self.training and w.ndim == 2:
```
Comment (152334H) [inline records/track_10min_16mb/2026-03-21_11L_XSA4_EMA_PartialRoPE_LateQAT_1.1248/train_gpt.py:540]: this gets [inlined as a ConstantVariable](https://github.com/pytorch/pytorch/blob/orig/release/2.10/torch/_dynamo/variables/user_defined.py#L319-L320); consequently late_qat's modification of ._qat_enabled doesn't work.

you can prove this by replacing the if-block with an `assert False`; it does not run

Comment (jfprincz) [reply records/track_10min_16mb/2026-03-21_11L_XSA4_EMA_PartialRoPE_LateQAT_1.1248/train_gpt.py:540]: thank you, i've updated the submission
